token.go

```diff
@@ -164,36 +166,125 @@ type memoryStore struct {
+	if err != nil {
+		return "", err
+	}
+
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.refreshToken
+	return s.AccessToken, err
 }
 
-func (s *memoryStore) ValidFor() time.Duration {
+func (s *memoryStore) GetRefreshToken() (string, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.validUntil.Sub(time.Now())
+	return s.refreshToken, nil
 }
 
-func (s *memoryStore) Update(r *TokenRefreshResponse) {
+func (s *persistentStore) GetRefreshToken() (string, error) {
+	err := s.getPersistentTokenData()
+	if err != nil {
+		return "", err
+	}
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.RefreshToken, err
+}
+
+func (s *memoryStore) GetValidFor() (time.Duration, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.validUntil.Sub(time.Now()), nil
+}
+
+func (s *persistentStore) GetValidFor() (time.Duration, error) {
+	err := s.getPersistentTokenData()
+	if err != nil {
+		return 0, err
+	}
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.ValidUntil.Sub(time.Now()), err
+}
+
+func (s *memoryStore) Update(r *TokenRefreshResponse) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	s.accessToken = r.AccessToken
 	s.refreshToken = r.RefreshToken
-	// Subtract a few seconds to allow for network and processing delays.
-	s.validUntil = time.Now().Add(r.ExpiresIn.Duration - (15 * time.Second))
+	s.validUntil = generateValidUntil(r)
+
+	return nil
 }
 
-// NewMemoryTokenStore is a TokenStore with no persistence.
-func NewMemoryTokenStore(r *TokenRefreshResponse) TokenStore {
+func (s *persistentStore) Update(r *TokenRefreshResponse) error {
+	f, err := os.Create("/tmp/tokenStore")
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Update in-memory data
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.AccessToken = r.AccessToken
+	s.RefreshToken = r.RefreshToken
+	s.ValidUntil = generateValidUntil(r)
+
+	// Write token data to file to be accessed later
+	encoder := gob.NewEncoder(f)
+	err = encoder.Encode(s)
+
+	return err
+}
+
+// NewMemoryTokenStore is a TokenStorer with no persistence.
+func NewMemoryTokenStore(r *TokenRefreshResponse) TokenStorer {
 	s := &memoryStore{}
 	s.Update(r)
 	return s
 }
+
+// NewPersistentTokenStore is a ToeknStorer with persistence to disk
+func NewPersistentTokenStore(r *TokenRefreshResponse) (TokenStorer, error) {
+	s := &persistentStore{}
+	// update persistent storage
+	if err := s.Update(r); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// generateValidUntil returns the time the token expires with an added buffer
+func generateValidUntil(r *TokenRefreshResponse) time.Time {
+	// Subtract a few seconds to allow for network and processing delays.
+	return time.Now().Add(r.ExpiresIn.Duration - (15 * time.Second))
+}
+
+// getPersistentTokenData returns the token data stored in a local file
+func (s *persistentStore) getPersistentTokenData() error {
+	f, err := os.Open("/tmp/tokenStore")
+	if err != nil {
+		return err
+	}
+	decoder := gob.NewDecoder(f)
+	err = decoder.Decode(s)
```

Comment:
This is operating on the `persistentStore` without locking `mu`, which is racy. How does `Decode` interact with the mutex member, anyway? Does it affect it at all?

Reply:
This should be addressed now.